downstairs/src/lib.rs

```diff
@@ -940,6 +941,11 @@ struct ConnectionState {
     /// Upstairs connection, populated at `HelloItsMe`
     upstairs_connection: Option<UpstairsConnection>,
 
+    // It's important for `cancel` to be above `reply_channel_tx`, to ensure
+    // that `recv_task` knows it has been cancelled by the time `reply_task`
+    // has stopped.  This lets us distinguish between "`reply_task` stopped due
+    // to an error" and "`reply_task` stopped because `reply_channel_tx` was
+    // dropped by the `Downstairs`".
```

Comment:
This feels important enough to `impl Drop for ConnectionState` and do it ourselves?

Reply:
I realized that we still need to be robust against ordering shenanigans, so I'm just going to remove the comment entirely, since it's not load-bearing anymore.

The specific ordering issue that _isn't_ prevented by careful `Drop` ordering is

- `recv_task` exits with an error
- Downstairs tries to write to it, notices that it's gone, removes that connection (which cancels the `reply_task` token)
- `reply_task` notices that `recv_task` exited, tells the Downstairs to remove the connection (before noticing that its token has been cancelled)